src/DMLQueryBuilder.php

```diff
@@ -74,42 +69,28 @@ public function upsert(
     ): string {
         $insertSql = $this->insert($table, $insertColumns, $params);
 
-        /** @psalm-var array $uniqueNames */
-        [$uniqueNames, , $updateNames] = $this->prepareUpsertColumns(
-            $table,
-            $insertColumns,
-            $updateColumns,
-        );
+        [$uniqueNames, , $updateNames] = $this->prepareUpsertColumns($table, $insertColumns, $updateColumns);
 
         if (empty($uniqueNames)) {
             return $insertSql;
         }
 
-        if ($updateNames === []) {
+        if ($updateColumns === false || $updateNames === []) {
```

Comment:
```suggestion
        if (empty($updateColumns)) {
```

It's a little more productive.

Reply:
Here is not only `$updateColumns === false` but also `$updateNames === []`

and `$updateNames` could be `null`